core/include/scipp/core/values_and_variances_strong_types.h

```diff
@@ -0,0 +1,202 @@
+  template <class... Args>
+  Dims(Args &&... args) : dims(std::forward<Args>(args)...) {}
+  template <class T>
+  Dims(std::initializer_list<T> init) : dims(init.begin(), init.end()) {}
+};
+
+namespace detail {
+
+template <class Tag, class... Ts> struct TaggedTuple {
+  using tag_type = Tag;
+  using tuple_type = std::tuple<Ts...>;
+  tag_type tag;
+  tuple_type tuple;
+};
+
+struct ValuesTag {};
+
+struct VariancesTag {};
+
+template <class Tag, class... Ts> auto makeTaggedTuple(Ts &&... ts) {
+  return TaggedTuple<Tag, std::remove_reference_t<Ts>...>{
+      {}, std::make_tuple(std::forward<Ts>(ts)...)};
+}
+
+template <class Tag, class T>
+auto makeTaggedTuple(std::initializer_list<T> init) {
+  using iter = typename std::initializer_list<T>::iterator;
+  return TaggedTuple<VariancesTag, iter, iter>{
+      {}, std::make_tuple(init.begin(), init.end())};
+}
+
+template <class Tag, class... Args>
+using MakeTaggedTupleReturnType =
+    decltype(makeTaggedTuple<Tag>(std::declval<Args>()...));
+} // namespace detail
+
+template <class... Args>
+using ValuesTaggedTupple =
+    detail::MakeTaggedTupleReturnType<detail::ValuesTag, Args...>;
+
+template <class... Args> struct Values : ValuesTaggedTupple<Args...> {
+  using tag_type = detail::ValuesTag;
+  Values(Args &&... args)
+      : ValuesTaggedTupple<Args...>(
+            detail::makeTaggedTuple<tag_type>(std::forward<Args>(args)...)) {}
+  template <class T>
+  Values(std::initializer_list<T> init)
+      : ValuesTaggedTupple<Args...>(
+            detail::makeTaggedTuple<tag_type>(init.begin(), init.end())) {}
+};
+template <class... Args> Values(Args &&... args)->Values<Args...>;
+template <class T>
+Values(std::initializer_list<T>)
+    ->Values<typename std::initializer_list<T>::iterator,
+             typename std::initializer_list<T>::iterator>;
+
+template <class... Args>
+using VariancesTaggedTupple =
+    detail::MakeTaggedTupleReturnType<detail::VariancesTag, Args...>;
+
+template <class... Args> struct Variances : VariancesTaggedTupple<Args...> {
+  using tag_type = detail::VariancesTag;
+  Variances(Args &&... args)
+      : VariancesTaggedTupple<Args...>(
+            detail::makeTaggedTuple<tag_type>(std::forward<Args>(args)...)) {}
+  template <class T>
+  Variances(std::initializer_list<T> init)
+      : VariancesTaggedTupple<Args...>(
+            detail::makeTaggedTuple<tag_type>(init.begin(), init.end())) {}
+};
+template <class... Args> Variances(Args &&... args)->Variances<Args...>;
+template <class T>
+Variances(std::initializer_list<T>)
+    ->Variances<typename std::initializer_list<T>::iterator,
+                typename std::initializer_list<T>::iterator>;
+
+namespace detail {
+template <class Tag, class T> struct has_tag : std::false_type {};
+
+template <class Tag, class... Ts>
+struct has_tag<Tag, Values<Ts...>> : std::is_same<Tag, ValuesTag> {};
+
+template <class Tag, class... Ts>
+struct has_tag<Tag, Variances<Ts...>> : std::is_same<Tag, VariancesTag> {};
```

Comment:
Better traits in here? Something to utilize `T::tag_type` instead of explicit specialization for particular classes.